spec/classes/server_spec.rb

```diff
@@ -0,0 +1,1311 @@
+#    },
+#    'Solaris-5.9' => {
+#      :architecture           => 'i86pc',
+#      :osfamily               => 'Solaris',
+#      :kernelrelease          => '5.9',
+#      :ssh_version            => 'Sun_SSH_2.2',
+#      :ssh_version_numeric    => '2.2',
+#      :ssh_packages           => ['SUNWsshcu', 'SUNWsshdr', 'SUNWsshdu', 'SUNWsshr', 'SUNWsshu'],
+#      :sshd_config_mode       => '0644',
+#      :sshd_service_name      => 'sshd',
+#      :sshd_service_hasstatus => false,
+#      :sshd_config_fixture    => 'sshd_config_solaris',
+#      :ssh_config_fixture     => 'ssh_config_solaris',
+#    },
+#    'Ubuntu-1604' => {
+#      :architecture           => 'x86_64',
+#      :osfamily               => 'Debian',
+#      :operatingsystemrelease => '16.04',
+#      :ssh_version            => 'OpenSSH_7.2p2',
+#      :ssh_version_numeric    => '7.2',
+#      :ssh_packages           => ['openssh-server', 'openssh-client'],
+#      :sshd_config_mode       => '0600',
+#      :sshd_service_name      => 'ssh',
+#      :sshd_service_hasstatus => true,
+#      :sshd_config_fixture    => 'sshd_config_ubuntu1604',
+#      :ssh_config_fixture     => 'ssh_config_ubuntu1604',
+#    },
+  }
+
+  defaults = {
+    :fqdn                   => 'monkey.example.com',
+    :hostname               => 'monkey',
+    :ipaddress              => '127.0.0.1',
+    :root_home              => '/root',
+    :specific               => 'dummy',
+    :sshrsakey              => 'AAAAB3NzaC1yc2EAAAABIwAAAQEArGElx46pD6NNnlxVaTbp0ZJMgBKCmbTCT3RaeCk0ZUJtQ8wkcwTtqIXmmiuFsynUT0DFSd8UIodnBOPqitimmooAVAiAi30TtJVzADfPScMiUnBJKZajIBkEMkwUcqsfh630jyBvLPE/kyQcxbEeGtbu1DG3monkeymanOBW1AKc5o+cJLXcInLnbowMG7NXzujT3BRYn/9s5vtT1V9cuZJs4XLRXQ50NluxJI7sVfRPVvQI9EMbTS4AFBXUej3yfgaLSV+nPZC/lmJ2gR4t/tKvMFF9m16f8IcZKK7o0rK7v81G/tREbOT5YhcKLK+0wBfR6RsmHzwy4EddZloyLQ==',
+  }
+
+  defaults_solaris = {
+    :fqdn                => 'monkey.example.com',
+    :hostname            => 'monkey',
+    :ipaddress           => '127.0.0.1',
+    :kernelrelease       => '5.10',
+    :osfamily            => 'Solaris',
+    :root_home           => '/root',
+    :specific            => 'dummy',
+    :ssh_version         => 'Sun_SSH_2.2',
+    :ssh_version_numeric => '2.2',
+    :sshrsakey           => 'AAAAB3NzaC1yc2EAAAABIwAAAQEArGElx46pD6NNnlxVaTbp0ZJMgBKCmbTCT3RaeCk0ZUJtQ8wkcwTtqIXmmiuFsynUT0DFSd8UIodnBOPqitimmooAVAiAi30TtJVzADfPScMiUnBJKZajIBkEMkwUcqsfh630jyBvLPE/kyQcxbEeGtbu1DG3monkeymanOBW1AKc5o+cJLXcInLnbowMG7NXzujT3BRYn/9s5vtT1V9cuZJs4XLRXQ50NluxJI7sVfRPVvQI9EMbTS4AFBXUej3yfgaLSV+nPZC/lmJ2gR4t/tKvMFF9m16f8IcZKK7o0rK7v81G/tREbOT5YhcKLK+0wBfR6RsmHzwy4EddZloyLQ==',
+  }
+
+  default_facts = osfamily_matrix['EL-7'].merge(defaults)
+
+  let(:facts) { default_facts }
+
+  osfamily_matrix.each do |os, facts|
+    context "with default params on osfamily #{os}" do
+      let(:facts) { defaults.merge(facts)}
+
+      # FIXME - first one fails. If you remove the duplicate, the first compile
```

Comment:
@Phil-Friderici I could really use your help on this one. I can't seem to debug why the first test always fails and why passing `~` as nil/undef from hiera data in module is not working as expected.

I've been rewriting this module for Puppet 5/6 and doing it against the man pages so it supports every option and has the appropriate type checking. It also has a `custom` param so you can add whatever else to the configs.

Reply:
This seems to have to due with running puppet as a gem under osx

Reply:
@ghoneycutt After `pdk convert` I don't get that error. I only get errors for the tests of ssh_version facts. Though `pdk validate` throws alot of errors. But that should be only some minor work.

If you are ok to move to PDK (which I really recommend) you can have a look at my PR:
https://github.com/Phil-Friderici/puppet-module-ssh/pull/8

Did convert some modules to PDK already , so I am happy to help with that. Please let me know :)

Reply:
The errors for the fact disapeared after using mocha for mocking :)

Reply:
Based on the v4 branch, I have fixed the unit tests. I have created a PR with the needed changes (https://github.com/ghoneycutt/puppet-module-ssh/pull/370).

Please review or comment :)